### ISSUE TITLE
test_tensor_with_grad_to_scalar_warning failure

### ISSUE THREAD

**Issue by rraminen** [pytorch/pytorch#157252]
### 🐛 Describe the bug

The test `test/test_torch.py::TestTorch::test_tensor_with_grad_to_scalar_warning FAILED [0.0010s] `fails when run as a part of test suite `test_torch.py`. But it passes when run stand alone.

The test `test/test_torch.py::TestTorch::test_tensor_with_grad_to_scalar_warning` is failing because of `test/test_torch.py::TestTorch::test_linspace_logspace` which is run prior to it in `test_torch.py` test suite.

PyTorch internally might have  trigger a one-time warning in `test_linspace_logspace`. That could be the same warning the other test (`test_tensor_with_grad_to_scalar_warning`) is trying to catch, but it's already been triggered and silenced in `test_linspace_logspace`.

That’s why len(w) becomes 0 or the warning isn’t present anymore. it’s already been raised once.

So calling factory functions like `torch.linspace`, `torch.logspace`, etc., inside a `with torch.no_grad():` block especially when testing gradients will solve this dependency issue.

```
# git diff  test/test_torch.py
diff --git a/test/test_torch.py b/test/test_torch.py
index 5094d96d3c..9858dc8d64 100644
--- a/test/test_torch.py
+++ b/test/test_torch.py
@@ -7007,22 +7007,23 @@ class TestTorch(TestCase):
         end = 3.0

         for step in [0, 1, 2]:
-            self.assertFalse(
-                torch.linspace(
-                    torch.tensor(start, requires_grad=True),
-                    torch.tensor(end, requires_grad=True), step
-                ).requires_grad
-            )
-            self.assertFalse(torch.linspace(torch.tensor(start, requires_grad=True), end, step).requires_grad)
-            self.assertFalse(torch.linspace(start, torch.tensor(end, requires_grad=True), step).requires_grad)
-            self.assertFalse(
-                torch.logspace(
-                    torch.tensor(start, requires_grad=True),
-                    torch.tensor(end, requires_grad=True), step
-                ).requires_grad
-            )
-            self.assertFalse(torch.logspace(torch.tensor(start, requires_grad=True), end, step).requires_grad)
-            self.assertFalse(torch.logspace(start, torch.tensor(end, requires_grad=True), step).requires_grad)
+            with torch.no_grad():
+                self.assertFalse(
+                    torch.linspace(
+                        torch.tensor(start, requires_grad=True),
+                        torch.tensor(end, requires_grad=True), step
+                    ).requires_grad
+                )
+                self.assertFalse(torch.linspace(torch.tensor(start, requires_grad=True), end, step).requires_grad)
+                self.assertFalse(torch.linspace(start, torch.tensor(end, requires_grad=True), step).requires_grad)
+                self.assertFalse(
+                    torch.logspace(
+                        torch.tensor(start, requires_grad=True),
+                        torch.tensor(end, requires_grad=True), step
+                    ).requires_grad
+                )
+                self.assertFalse(torch.logspace(torch.tensor(start, requires_grad=True), end, step).requires_grad)
+                self.assertFalse(torch.logspace(start, torch.tensor(end, requires_grad=True), step).requires_grad)

     # FIXME: move to shape ops test suite
```

### Versions

None

**Comment by soulitzer** [comment]
Thanks for the investigation! We recently made this test run under`set_warn_always_context` so even if it has been triggered for the same process it will always fire, so I think it should be fixed now.
